lang/src/bpf_writer.rs

```diff
@@ -0,0 +1,28 @@
+use solana_program::program_memory::sol_memcpy;
+use std::cmp;
+use std::io::{self, Write};
+
+#[derive(Debug, Default)]
+pub struct BpfWriter<T> {
+    inner: T,
+    pos: u64,
+}
+
+impl<T> BpfWriter<T> {
+    pub fn new(inner: T) -> Self {
+        Self { inner, pos: 0 }
+    }
+}
+
+impl Write for BpfWriter<&mut [u8]> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let pos = cmp::min(self.pos, self.inner.len() as u64);
+        sol_memcpy(&mut self.inner[(pos as usize)..], buf, buf.len());
+        self.pos += buf.len() as u64;
+        Ok(buf.len())
+    }
```

Comment:
I think @armaniferrante should also look at this but I think that this implementation is not equivalent to the Cursor one. In the `Cursor` impl, there will be no write if `self.pos >= self.inner.len()` (because the slice[pos..] will be empty which will make the `Slice::write` call use an `amt` of 0). however, `sol_memcpy` just copies the data in `C` with pointer arithmetic so the buffer will _always_ be completely written, even if that means it's written to an unintended mem region.

I think the following would work
```
// pseudo
if pos >= self.inner.len() -> return 0
amt = self.inner.len() - pos
sol_memcpy(&mut self.inner[(amt as usize)..], buf, amt);
Ok(amt)
```
and then we implement `fn write_all`:
```
fn write_all(&mut self, data: &[u8]) -> io::Result<()> {
    if self.write(data)? == data.len() {
        Ok(())
    } else {
        Err(Error::new_const(ErrorKind::WriteZero, &"failed to write whole buffer"))
    }
}
```

Reply:
I had a similar thought today while i was on a plane but didn't have internet to look up to docs lol

Reply:
new commit resolves other conversations and implements this suggestion, but keep this one open for further discussion if needed.

Reply:
I guess with this change, there isn't really a need for the `pos` field anymore right?

Reply:
the `pos` is still needed. we write the discriminator with `write_all` and then serialize the account. my pseudocode above shouldve updated the pos though

Reply:
hmmm looks like its having some serialization issues with accounts in certain tests 🤔 trying to figure out what is wrong about this

Reply:
the writer writes `amt` which is all of the inner slice from pos onwards. if buf.len >= amt thats no problem but if buf.len() < amt, it is. so we should not write amt but cmp::min(buf.len(), amt)

Reply:
yep, that makes a lot of sense. seems good now